### PR TITLE
Indirect - Chi squared in Result workspace and plot combobox

### LINKED ISSUE #24111: Indirect - Chi-Squared should be in the result workspace and Plot ComboBox
**Original reporter:** Spencer

### Expected behavior
Chi-Squared should be in the result workspace and Plot ComboBox for Indirect Data Analysis.

### Steps to reproduce the behavior
1.`Interfaces`->`Indirect`->`Data Analysis`
2. Run a fit on MSDFit, IqtFit, ConvFit or F(Q) Fit using the data below
3. The result workspaces last row should be Chi-Squared
4. An option to plot chi-squared should be given in the `Plot` Combobox

### Platforms affected
All

### PR DESCRIPTION
**Pull request by robertapplin**
**Description of work.**
This PR ensures that Chi_squared is added to the bottom row of the Results workspaces for Indirect Data Analysis (MSDFit, IqtFit, ConvFit and F(Q) Fit).
This PR also allows the plotting of Chi_squared in the output options.

**To test:**
1.`Interfaces`->`Indirect`->`Data Analysis`
2. In turn, run a fit on MSDFit, IqtFit, ConvFit and F(Q) Fit using the data below
3. The _Result workspace's last row should be Chi-Squared
4. An option to plot chi-squared should be given in the output options in the `Plot` Combobox
5. This should plot the values that can be found in the result workspace

**Data**
*MSDFit:*
[osi104367-104371_graphite002_red_elwin_eq.zip](https://github.com/mantidproject/mantid/files/2910130/osi104367-104371_graphite002_red_elwin_eq.zip)

*IqtFit:*
[irs26176_graphite002_iqt.zip](https://github.com/mantidproject/mantid/files/2910139/irs26176_graphite002_iqt.zip)

*ConvFit:*
[irs26176_graphite002_red.zip](https://github.com/mantidproject/mantid/files/2910143/irs26176_graphite002_red.zip)
[irs26173_graphite002_res.zip](https://github.com/mantidproject/mantid/files/2910144/irs26173_graphite002_res.zip)

*FQFit:*
[IN16B_125878_QLd_Result.zip](https://github.com/mantidproject/mantid/files/2910147/IN16B_125878_QLd_Result.zip)

The plotting options should be disbaled for F(Q) Fit because there is only one data point. Look in the `IN16B_125878_QLd_Result_HWHM_FofQFit_ChudleyElliot` workspace for chi_squared.

Fixes #24111 

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/IndividualTicketTesting/)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/WorkflowAlgorithms/src/ProcessIndirectFitParameters.cpp [hunk 1]
```diff
@@ -127,6 +127,12 @@ std::vector<Column_const_sptr> extractColumns(ITableWorkspace const *table,
   return columns;
 }
 
+void populateChiSquaredErrorsWithZeros(std::vector<double> &e,
+                                       std::size_t const &difference) {
+  for (auto i = 0u; i < difference; ++i)
+    e.emplace_back(0.0);
```
Comment (soininen) [inline Framework/WorkflowAlgorithms/src/ProcessIndirectFitParameters.cpp]: `std::fill` or `std::fill_n` should do the job as well. I am unsure if you need this entire function just for this purpose.

File: Framework/WorkflowAlgorithms/src/ProcessIndirectFitParameters.cpp [hunk 2]
```diff
@@ -150,6 +157,9 @@ struct TableToMatrixWorkspaceConverter {
     extractValuesFromColumns<double>(startRow, endRow, m_eColumns,
                                      std::back_inserter(e));
 
+    if (includeChiSquared)
+      populateChiSquaredErrorsWithZeros(e, y.size() - e.size());
```
Comment (soininen) [inline Framework/WorkflowAlgorithms/src/ProcessIndirectFitParameters.cpp]: `std::fill_n`

**Review by soininen — CHANGES_REQUESTED** [review]
Chi squared is added to the result workspaces and can be plotted as described. The code changes look good too. However, I propose to use `std::fill_n` as explained in the comments below. Two issues I encountered which are not in the scope of this PR but perhaps could be tackled later:

- The result workspace when choosing the 'Fit' option from the 'Fit' combo box-button does not have the chi squared row. I don't know if this is desired but it is inconsistent.
- When setting the plot panel as floating, there is no way of moving the panel on my system (Ubuntu 16.04). I can only resize it in the floating state.

**Comment by SpencerHowells** [comment]
@soininen @robertapplin Regarding the floating panel. Marina discovered that problem during testing. It is a bug in Qt that affects linux only.

**Comment by robertapplin** [comment]
@soininen Thank you for finding these bugs. As @SpencerHowells said the floating panel not moving on linux is a qt bug. Regarding the Fit ComboBox, this combobox will be removed when Simultaneous fitting has been implemented and so it is probably not worth looking into this

**Comment by soininen** [comment]
There seems to be a merge conflict now.

**Comment by robertapplin** [comment]
Thanks @soininen , the conflict should be fixed now.

**Review by soininen — APPROVED** [review]

**Pull request merged**